compiler/rustc_ast/src/ast.rs

```diff
@@ -1092,15 +1092,9 @@ impl Expr {
         if let ExprKind::Block(ref block, _) = self.kind {
             match block.stmts.last().map(|last_stmt| &last_stmt.kind) {
                 // Implicit return
-                Some(&StmtKind::Expr(_)) => true,
-                Some(&StmtKind::Semi(ref expr)) => {
-                    if let ExprKind::Ret(_) = expr.kind {
-                        // Last statement is explicit return.
-                        true
-                    } else {
-                        false
-                    }
-                }
+                Some(StmtKind::Expr(_)) => true,
+                // Last statement is an explicit return?
+                Some(StmtKind::Semi(expr)) => matches!(expr.kind, ExprKind::Ret(_)),
```

Comment:
```suggestion
                Some(StmtKind::Expr(_) | StmtKind::Semi(Expr { kind: ExprKind::Ret(_), .. })) => true,
```

I wonder if it is possible to do this?

Reply:
it is, but it's very noisy. I think we're waiting for pattern aliases before starting to merge nested matches like that.

Reply:
Waiting for pattern aliases?

Reply:
Well... once we have a way to create pattern aliases in a more convenient way than what `macro_rules` macros can offer, then we can create a pattern alias that allows us to conveniently match on an `Expr` without having to specify all the `kind: Foo, ..` boilerplate.

Reply:
Having it split up also gives a good place for the helpful comments on each branch.

Reply:
How about this?
```suggestion
                Some(StmtKind::Expr(_)) => true,
                Some(StmtKind::Semi(Expr { kind: ExprKind::Ret(_), .. })) => true,
```

Reply:
@oli-obk This is the first I've heard about pattern aliases as a wanted feature (other than from me)!  Is there an issue/discussion already?